synapse/storage/databases/main/event_federation.py

```diff
@@ -151,15 +152,179 @@ async def get_auth_chain_difference(
+            )
+            txn.execute(sql % (clause,), args)
+
+            for event_id, chain_id, sequence_number in txn:
+                chain_info[event_id] = (chain_id, sequence_number)
+                seen_chains.add(chain_id)
+                chain_to_event.setdefault(chain_id, {})[sequence_number] = event_id
+
+        # Corresponds to `state_sets`, except as a map from chain ID to max
+        # sequence number reachable from the state set.
+        set_to_chain = []  # type: List[Dict[int, int]]
+        for state_set in state_sets:
+            chains = {}  # type: Dict[int, int]
+            set_to_chain.append(chains)
+
+            for event_id in state_set:
+                chain_id, seq_no = chain_info[event_id]
+
+                curr = chains.setdefault(chain_id, seq_no)
+                if curr < seq_no:
+                    chains[chain_id] = seq_no
+
+        # Now we lok up all links for the chains we have, adding chains to
+        # set_to_chain that are reachable from each set.
+        sql = """
+            SELECT
+                origin_chain_id, origin_sequence_number,
+                target_chain_id, target_sequence_number
+            FROM event_auth_chain_links
+            WHERE %s
+        """
+
+        for batch in batch_iter(seen_chains, 1000):
+            clause, args = make_in_list_sql_clause(
+                txn.database_engine, "origin_chain_id", batch
+            )
+            txn.execute(sql % (clause,), args)
+
+            for (
+                origin_chain_id,
+                origin_sequence_number,
+                target_chain_id,
+                target_sequence_number,
+            ) in txn:
+                for chains in set_to_chain:
+                    # chains are only reachable if the origin sequence number of
+                    # the link is less than the max sequence number in the
+                    # origin chain.
+                    if origin_sequence_number <= chains.get(origin_chain_id, 0):
+                        curr = chains.setdefault(
+                            target_chain_id, target_sequence_number
+                        )
+                        if curr < target_sequence_number:
+                            chains[target_chain_id] = target_sequence_number
+
+                seen_chains.add(target_chain_id)
+
+        # Now for each chain we figure out the maximum sequence number reachable
+        # from *any* state set and the minimum sequence number reachable from
+        # *all* state sets. Events in that range are in the auth chain
+        # difference.
+        result = set()
+
+        chain_to_gap = {}
+        for chain_id in seen_chains:
+            min_seq_no = min(chains.get(chain_id, 0) for chains in set_to_chain)
+
+            max_seq_no = 0
+            for chains in set_to_chain:
+                s = chains.get(chain_id)
+                if s:
+                    max_seq_no = max(max_seq_no, s)
```

Comment:
I'm struggling to figure out how this loop is different than `max_seq_no = max(chains.get(chain_id, 0) for chains in set_to_chain)` (essentially the same as above).

Reply:
I.. don't think it is in the end. I guess it probably was different at some point, :shrug: